docs/abi.rst

```diff
@@ -0,0 +1,652 @@
+:any:`abi.AssetTransferTransaction` :code:`appl`           Static           An application call transaction
+=================================== ====================== ================ =======================================================================================================================================================
+
+Usage
+""""""""""""""""""""""""""""""""""""""""""
+
+Getting the Transaction Group Index
+''''''''''''''''''''''''''''''''''''
+
+All transaction types implement the :any:`abi.Transaction.index()` method, which returns the absolute index of that transaction in the group.
+
+A brief example is below:
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def handle_txn_args(
+        any_txn: abi.Transaction,
+        pay: abi.PaymentTransaction,
+        axfer: abi.AssetTransferTransaction,
+    ) -> Expr:
+        return Seq(
+            Assert(any_txn.index() == Txn.group_index() - Int(3)),
+            Assert(pay.index() == Txn.group_index() - Int(2)),
+            Assert(axfer.index() == Txn.group_index() - Int(1)),
+        )
+
+Accessing Transaction Fields
+'''''''''''''''''''''''''''''
+
+All transaction types implement the :any:`abi.Transaction.get()` method, which returns a :any:`TxnObject` instance that can be used to access fields from that transaction.
+
+A brief example is below:
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def deposit(payment: abi.PaymentTransaction, sender: abi.Account) -> Expr:
+        """This method receives a payment from an account opted into this app and records it in their
+        local state.
+        """
+        return Seq(
+            Assert(payment.get().sender() == sender.address()),
+            Assert(payment.get().receiver() == Global.current_application_address()),
+            App.localPut(
+                sender.address(),
+                Bytes("balance"),
+                App.localGet(sender.address(), Bytes("balance")) + payment.get().amount(),
+            ),
+        )
+
+Limitations
+""""""""""""""""""""""""""""""""""""""""""
+
+TODO: explain limitations, such as can't be created directly, used as method return value, or embedded in other types
+
+
+Subroutines with ABI Types
+--------------------------
+
+Subroutines can be created that accept ABI types are arguments and produce ABI types as return values. PyTeal will type check all subroutine calls and ensure that the correct types are being passed to such subroutines and that their return values are used correctly.
+
+There are two different ways to use ABI types in subroutines, depending on whether the return value is an ABI type or a PyTeal :code:`Expr`.
+
+Subroutines that Return Expressions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you'd like to create a subroutine that accepts some or all arguments as ABI types, but whose return value is a PyTeal :code:`Expr`, the normal :any:`@Subroutine <Subroutine>` decorator can be used.
```

Comment:
@jasonpaulos Minor and optional:  I think _normal_ can be removed because the two discussed subroutine decorators have unique names.  

```suggestion
If you'd like to create a subroutine that accepts some or all arguments as ABI types, but whose return value is a PyTeal :code:`Expr`, the :any:`@Subroutine <Subroutine>` decorator can be used.
```